optee-utee/src/property.rs

```diff
@@ -0,0 +1,481 @@
+        let mut raw_uuid = raw::TEE_UUID {
+            timeLow: 0,
+            timeMid: 0,
+            timeHiAndVersion: 0,
+            clockSeqAndNode: [0; 8],
+        };
+
+        let res =
+            unsafe { raw::TEE_GetPropertyAsUUID(set, key.as_ptr() as *const core::ffi::c_char, &mut raw_uuid) };
+        if res != 0 {
+            return Err(Error::from_raw_error(res));
+        }
+
+        Ok(Uuid::from_raw(raw_uuid))
+    }
+}
+
+impl PropertyValue for Identity {
+    fn from_raw(set: raw::TEE_PropSetHandle, key: CString) -> Result<Self> {
+        // Allocate a buffer for the raw identity
+        let mut raw_id = raw::TEE_Identity {
+            login: 0,
+            uuid: raw::TEE_UUID {
+                timeLow: 0,
+                timeMid: 0,
+                timeHiAndVersion: 0,
+                clockSeqAndNode: [0; 8],
+            },
+        };
+
+        let res = unsafe {
+            raw::TEE_GetPropertyAsIdentity(set, key.as_ptr() as *const core::ffi::c_char, &mut raw_id)
+        };
+        if res != 0 {
+            return Err(Error::from_raw_error(res));
+        }
+
+        Ok(Identity::from_raw(raw_id))
+    }
+}
+
+/// Represents a TEE property key.
+/// The property key is used to identify a specific property
+/// within a property set. The property key is a string that
+/// is used to query the property value from the TEE property
+/// set. The property key is defined in the TEE Internal API,
+/// such as "gpd.client.identity" or "gpd.tee.apiversion".
+pub trait PropertyKey {
+    type Output: PropertyValue;
+    fn key(&self) -> CString;
+    fn set(&self) -> PropertySet;
+
+    fn get(&self) -> Result<Self::Output> {
+        Self::Output::from_raw(self.set().as_raw(), self.key())
+    }
+}
+
+/// Macro to define a property key.
+/// This macro generates a struct that implements the
+/// PropertyKey trait.
+macro_rules! define_property_key {
+    (
+        $name:ident,
+        $set:ident,
+        $key:literal,
+        $output:ty
+    ) => {
+        pub struct $name;
+
+        impl PropertyKey for $name {
+            type Output = $output;
+
+            fn key(&self) -> CString {
+                CString::new($key).unwrap_or_default()
+            }
+
+            fn set(&self) -> PropertySet {
+                PropertySet::$set
+            }
+        }
+    };
+}
+
+// Define all existing property keys for the TEE property set.
+// The format is:
+// `define_property_key!(Name, Set, "key", OutputType);`
+// The `Set` is one of the PropertySet it belongs to.
+// The `key` is the raw property key string.
+// The `OutputType` is the type of the property value.
+// 
+// To get the property value, use the `get` method.
+// Example usage:
+// 
+// ``` no_run
+// use optee_utee::{PropertyKey, CurrentTaAppId};
+// 
+// let my_property = CurrentTaAppId.get()?;
+// ```
+define_property_key!(CurrentTaAppId, CurrentTa, "gpd.ta.appID", Uuid);
```

Comment:
The idea here is to abstract the complexity of fetching properties under Property Sets by introducing dedicated types. The main concern is the naming convention used for these newly introduced types.

The suggested approach is to map type names directly from the corresponding property keys (or their suffixes). For example:

gpd.ta.doesNotCloseHandleOnCorruptObject → TaDoesNotCloseHandleOnCorruptObject

gpd.tee.apiversion → TeeApiVersion

gpd.ta.dataSize → TaDataSize

This naming convention allows developers to intuitively derive the type name from the GlobalPlatform documentation without needing to inspect the implementation details. It significantly improves developer experience and code readability.